lib/changes.go

```diff
@@ -18,30 +18,31 @@ type ChangeReport = changes.ChangeReportResponse
 
 // ChangeReport resolves the requested datasets and tries to generate a change report
 func (m *DatasetMethods) ChangeReport(ctx context.Context, p *ChangeReportParams) (*ChangeReport, error) {
-	reportSource := ""
-
-	if m.inst.http != nil {
-		res := &ChangeReport{}
-		err := m.inst.http.Call(ctx, AEChanges, p, res)
-		if err != nil {
-			return nil, err
-		}
-		return res, nil
+	got, _, err := m.inst.Dispatch(ctx, dispatchMethodName(m, "changereport"), p)
+	if res, ok := got.(*ChangeReport); ok {
+		return res, err
 	}
+	return nil, dispatchReturnError(got, err)
+}
+
+// ChangeReport generates report of changes between two datasets
+func (datasetImpl) ChangeReport(scope scope, p *ChangeReportParams) (*ChangeReport, error) {
+	ctx := scope.Context()
+	reportSource := ""
 
-	right, _, err := m.inst.ParseAndResolveRef(ctx, p.RightRefstr, reportSource)
+	right, _, err := scope.ParseAndResolveRef(ctx, p.RightRefstr, reportSource)
 	if err != nil {
 		return nil, err
 	}
 
 	var left dsref.Ref
 	if p.LeftRefstr != "" {
-		if left, _, err = m.inst.ParseAndResolveRef(ctx, p.LeftRefstr, reportSource); err != nil {
+		if left, _, err = scope.ParseAndResolveRef(ctx, p.LeftRefstr, reportSource); err != nil {
 			return nil, err
 		}
 	} else {
 		left = dsref.Ref{Username: right.Username, Name: right.Name}
 	}
 
-	return changes.New(m.inst, m.inst.stats).Report(ctx, left, right, reportSource)
+	return changes.New(scope.inst, scope.inst.stats).Report(ctx, left, right, reportSource)
```

Comment:
I'd strongly prefer that users of `scope` don't directly access the `Instance` it holds onto.

For the first parameter, could you add a method to `scope` called `Loader` that returns `scope.inst`. There's already a method on `scope` with that name (my fault, sorry), please rename that older method to `ParseResolveFunc()` and update it's caller - I believe there is only one for `apply`.

For the second parameter, please add a method to `scope` called `Stats` that simply returns the stats component.

Reply:
*chefs kiss*